In [33]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score

df = pd.read_csv("wbc.csv")

### 1. Pisahkan antara variabel yang dapat digunakan dan variabel yang tidak dapat digunakan

In [34]:
df = df.drop(columns=['id', 'Unnamed: 32'])
X = df.drop(columns=['diagnosis'])


### 2. Proses encoding pada kolom "diagnosis".

In [35]:
le = LabelEncoder()
y = le.fit_transform(df['diagnosis'])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

### Soal 3.standardisasi, 4.seleksi fitur, 5.proses pengujian, dan 6. boleh pakai pipeline

In [37]:

pipe = Pipeline([
    ("scaler", StandardScaler()),                     # Jawaban Soal 3: Standardisasi
    ("sel", SelectKBest(score_func=f_classif, k=5)),  # Jawaban Soal 4: Seleksi Fitur
    ("clf", LogisticRegression(max_iter=1000))        # Jawaban Soal 5: Logistic Regression
])

# jawaban no 6 termasuk disini karna hanya peryataan yang boleh menggunakan pipeline

### 7. Berdasarkan hasil analisa Anda, berapa jumlah fitur terbaik yang dapat digunakan? Apa saja fitur tersebut?

In [39]:
best_k = 0
best_score = 0

# Looping untuk mencari jumlah k (fitur) terbaik
for k in range(1, len(X.columns) + 1):
    pipe.set_params(sel__k=k) # mengubah nilai k pada pipeline
    pipe.fit(X_train, y_train)
    pred = pipe.predict(X_test)
    score = accuracy_score(y_test, pred)

    # ketika akurasi lebih baik, maka nilai k akan tersimpan
    if score > best_score:
        best_score = score
        best_k = k

# Latih ulang pipeline menggunakan k terbaik untuk melihat fiturnya
pipe.set_params(sel__k=best_k)
pipe.fit(X_train, y_train)

# Jawaban: Berapa jumlah fitur terbaik?
print(f"Berdasarkan analisa, jumlah fitur terbaik adalah: {best_k} fitur (Akurasi: {best_score:.4f})\n")

# Jawaban: Apa saja fitur tersebut? (Diekstrak menggunakan syntax seperti di modul)
feat_names = X.columns
sel_step = pipe.named_steps["sel"]
mask = sel_step.get_support()
selected_names = feat_names[mask]
selected_scores = sel_step.scores_[mask]

# Mengurutkan fitur dari skor ANOVA tertinggi
top_features = sorted(zip(selected_names, selected_scores), key=lambda t: t[1], reverse=True)

print(f"Daftar {best_k} fitur terbaik tersebut adalah:")
for name, score in top_features:
    print(f"- {name} (Score ANOVA: {score:.2f})")

Berdasarkan analisa, jumlah fitur terbaik adalah: 14 fitur (Akurasi: 0.9825)

Daftar 14 fitur terbaik tersebut adalah:
- concave points_worst (Score ANOVA: 733.72)
- perimeter_worst (Score ANOVA: 717.25)
- radius_worst (Score ANOVA: 692.86)
- concave points_mean (Score ANOVA: 684.53)
- perimeter_mean (Score ANOVA: 548.41)
- area_worst (Score ANOVA: 522.19)
- radius_mean (Score ANOVA: 511.27)
- area_mean (Score ANOVA: 444.86)
- concavity_mean (Score ANOVA: 397.59)
- concavity_worst (Score ANOVA: 319.51)
- compactness_mean (Score ANOVA: 263.56)
- compactness_worst (Score ANOVA: 238.20)
- radius_se (Score ANOVA: 205.43)
- perimeter_se (Score ANOVA: 193.17)


jadi dari hasil analisis saya, ada 14 fitur terbaik. yakni :
- concave points_worst (Score ANOVA: 733.72)
- perimeter_worst (Score ANOVA: 717.25)
- radius_worst (Score ANOVA: 692.86)
- concave points_mean (Score ANOVA: 684.53)
- perimeter_mean (Score ANOVA: 548.41)
- area_worst (Score ANOVA: 522.19)
- radius_mean (Score ANOVA: 511.27)
- area_mean (Score ANOVA: 444.86)
- concavity_mean (Score ANOVA: 397.59)
- concavity_worst (Score ANOVA: 319.51)
- compactness_mean (Score ANOVA: 263.56)
- compactness_worst (Score ANOVA: 238.20)
- radius_se (Score ANOVA: 205.43)
- perimeter_se (Score ANOVA: 193.17)